storage/remote/queue_manager.go

```diff
@@ -359,62 +357,62 @@ func (t *QueueManager) reshardLoop() {
 
 func (t *QueueManager) reshard(n int) {
 	numShards.WithLabelValues(t.queueName).Set(float64(n))
-
-	t.shardsMtx.Lock()
 	newShards := t.newShards(n)
 	oldShards := t.shards
 	t.shards = newShards
-	t.shardsMtx.Unlock()
 
 	oldShards.stop(t.flushDeadline)
 
-	// We start the newShards after we have stopped (the therefore completely
-	// flushed) the oldShards, to guarantee we only every deliver samples in
+	// We start the newShards (worker) after we have stopped (the therefore completely
+	// flushed) the oldShards (worker), to guarantee we only every deliver samples in
 	// order.
 	newShards.start()
 }
 
 type shards struct {
 	qm      *QueueManager
-	queues  []chan *model.Sample
+	queue   chan *model.Sample
```

Comment:
By removing the per-shard queue, you also break the guarantee that samples for the same timeseries will be delivered in order.

Reply:
Yes, I also found this problem after testing. But I have no idea to solve it. Do you have some suggestion?

Reply:
> I also found this problem after testing

What problem?